nengo_gui/server.py

```diff
@@ -1,171 +1,511 @@
 
-        page = self.server.gui.create_page(filename, reset_cfg=reset_cfg)
+        try:
+            origin = urlparse(self.headers['Origin'])
+            assert origin.netloc.lower() in valid_srv_addrs
+        except KeyError:
+            raise Forbidden()
+        except AssertionError:
+            raise Forbidden()
+
+        try:
+            host = self.headers['Host'].lower()
+            assert host in valid_srv_addrs
+            key = self.headers['Sec-WebSocket-Key']
+            assert len(base64.b64decode(key)) == 16
+        except KeyError:
+            raise BadRequest()
+        except AssertionError:
+            raise BadRequest()
+
+        if self.headers['Sec-WebSocket-Version'] != '13':
+            raise UpgradeRequired(['Sec-WebSocket-Version: 13'])
+
+        sec = base64.b64encode(hashlib.sha1(
+            (key + WS_MAGIC).encode('ascii')).digest()).decode('ascii')
+        _sendall(self.request, response.format(sec=sec).encode('utf-8'))
+
+        self.ws = self.server.create_websocket(self.request)
+        self.ws.set_blocking(False)
+
+        command = self._get_command(self.ws_commands, self.resource)
+        if command is None:
+            self.ws_default()
+        else:
+            getattr(self, command)()
+        self.ws.close()
 
-        # read the template for the main page
-        html = pkgutil.get_data('nengo_gui', 'templates/page.html')
-        if isinstance(html, bytes):
-            html = html.decode("utf-8")
+    def ws_default(self):
+        raise InvalidResource(self.path)
 
-        # fill in the javascript needed and return the complete page
-        components = page.create_javascript()
-        return html % dict(components=components)
+    @classmethod
+    def _get_command(cls, commands, path):
+        if not path.startswith('/'):
+            path = '/' + path
 
-    def swi_shutdown(self, *path):
-        self.stop()
-        return "Shutting down..."
+        while len(path) > 0:
+            if path in commands:
+                return commands[path]
+            path = path.rsplit('/', 1)[0]
+        if '/' in commands:
+            return commands['/']
+        return None
 
-    def ws_viz_component(self, client, uid):
-        """Handles ws://host:port/viz_component with a websocket"""
-        # figure out what component is being connected to
 
-        gui = self.server.gui
+class WebSocket(object):
+    ST_OPEN, ST_CLOSING, ST_CLOSED = range(3)
 
-        component = gui.component_uids[int(uid)]
-        while True:
+    def __init__(self, socket):
+        self.socket = socket
+        self._buf = bytearray([])
+        self.state = self.ST_OPEN
+
+    def set_timeout(self, timeout):
+        self.socket.settimeout(timeout)
+
+    def set_blocking(self, flag):
+        self.socket.setblocking(flag)
+
+    def _read(self):
+        try:
+            self._buf = self._buf + bytearray(self.socket.recv(512))
+        except ssl.SSLWantReadError:
```

Comment:
It looks like it's just ```SSLError```, with an ```errno == 2``` to check for that particular error:

https://github.com/slackhq/python-slackclient/pull/30/commits/a1428a30b7bf5d63b3217ed07995f5944e66caf5